src/array.jl

```diff
@@ -339,6 +339,14 @@ end
     @inbounds A.refs[I...] = get!(A.pool, convert(T, v))
 end
 
+@inline function promote_rule{T,N,R,T2}(::Type{CatArray{T,N,R}}, ::Type{Array{T2}})
```

Comment:
You don'tneed `@inline`. Also, you should add a type parameter `<:CatArray` and promote to it instead of `CatArray` (which is a union type). Instead of `T` and  `T2`, the rest of the code base uses `S`  and `T`.